enhancements/subscription-content/subscription-content-access.md

```diff
@@ -0,0 +1,242 @@
+will fall to the Subscription Content Webhook. This mutating admission webhook and associated
+operator will do the following:
+
+Define an annotation that can be added to any pod to consume subscription content. The value for
+this annotation will determine which RHEL content will be available. Ex: annotating
+`subscription.openshift.io/version = rhel7` will make RHEL7 content available to the pod’s
+containers. For an annotated pod, the webhook will add the volume and volume mounts to all
+containers such that RHEL content can be consumed The operator for this webhook will create the
+projected-resource CSI driver objects that point to the entitlement keys and yum repo definitions
+for the cluster. The operator will also create RBAC roles that grant permission to the entitlement
+keys and yum repo definitions across namespaces. The keys and repo definitions are assumed to exist
+in the `openshift-config` namespace. These roles will be aggregated to the standard “edit” and
+“admin” roles for the cluster.
+
+The full details of this component will be specified in a future enhancement proposal.
+
+#### Extensions for OpenShift Builds
+
+OpenShift builds will be enhanced to support a narrow set of annotations, which are then passed
+through to the build pod. The annotation for the subscription content webhook will be one of the
+annotations included in the allowed list - others may also be considered.
+
+### Risks and Mitigations
+
+**Risk:** Sensitive information can be leaked to unauthorized users.
+
+_Mitigations:_ The proposal for the Projected-resource CSI driver includes RBAC protections and
+SELinux controls to ensure secrets are protected.
+
+**Risk:** Allowing arbitrary annotations in builds can lead to privilege escalations.
+
+_Mitigation:_ Builds will use an allowlist to block arbitrary annotations from being passed through
+to build pods.
+
+## Design Details
+
+### Test Plan
+
+N/A - each individual component will have their own test plan.
+
+### Graduation Criteria
+
+The following is a proposed release timeline:
+
+1. The Projected-resource CSI driver needs to reach maturity via a Tech Preview -> GA cycle.
+2. The subscription content webhook will explicitly depend on the Projected-resource CSI driver.
+   It’s lifecycle is tied down wrt Tech Preview -> GA Builds will require 2 and 3 in order to pass
+   subscription content through.
+3. The cluster subscription manager can be released as an independent component on its own cadence.
+   However, it should employ the conventions of the subscription content webhook to ensure seamless
+   behavior.
+
+### Upgrade / Downgrade Strategy
+
+Generally speaking, each of the components in this meta-enhancement will be add-ons, deliverable via
+OLM. That said, a case can be made that these should be added to the core OpenShift payload.
+
+### Version Skew Strategy
+
+Each component will be responsible for managing version skew during upgrade.
```

Comment:
Worth discussing the behavior if the customer is using rhel 7 hosts in a cluster.

Worth discussing how an existing cluster will get the secret when upgrading.  Will we alert in absence of secret?  What was thinking around that?

Reply:
This could be interesting, because if a RHEL 7 host is subscribed independently of the cluster, the containers on that host already have access to entitlements. The defaults in cri-o pass those through to containers (via `mounts.conf`). In this scenario, "do nothing" means your container could consume the host entitlement, but adding the annotation lets you consume a different entitlement with potentially different access.

Reply:
@derekwaynecarr I added this as a "Risk" - we could address by having the Subscription Injection operator apply a default annotation to every pod.

Reply:
Updated. However, adding the annotation to every pod introduces a worse risk of pods being denied. The current proposal for the injection operator+webhook will use a custom resource to expose the entitlement, and a SAR checks against the custom resource to ensure the requesting service account has permission to read the entitlement.

Is there an easy way to make a custom resource readable to every service account?

Reply:
Maybe aggregate it to the view role, but the view role is not bound to every SA by default IIRC.

So we are talking an operator that is monitoring namespace creation a la what is done with the builder SA, etc.

At most it would be something a cluster admin would opt into.

Whether that is "easy" ....

Reply:
Per @dmage - we can address this by granting read permissions to the `system:serviceaccount` group.